flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+
+  def apply(): KMeans = {
+    new KMeans()
+  }
+
+  // ========================================== Operations =========================================
+
+  /** Provides the operation that makes the predictions for individual examples.
+    * The label of the vector will be the index of the cluster the input vector belongs to.
+    *
+    * @tparam T
+    * @return A PredictOperation, through which it is possible to predict a value, given a
+    *         feature vector
+    */
+  implicit def predictVectors[T <: Vector] = {
+    new PredictOperation[KMeans, Seq[LabeledVector], T, Double](){
+
+      override def getModel(
+          self: KMeans,
+          predictParameters: ParameterMap)
+        : DataSet[Seq[LabeledVector]] = {
+
+        self.centroids match {
+          case Some(model) => model
+          case None => {
+            throw new RuntimeException("The KMeans model has not been trained. Call first fit" +
+              "before calling the predict operation.")
+          }
+        }
+      }
+
+      override def predict(value: T, model: Seq[LabeledVector]): Double = {
+        findNearestCentroid(value, model)._1
+      }
+    }
+  }
+
+  /**
+   * [[FitOperation]] which iteratively computes centroids that match the given input DataSet by
+   * adjusting the given initial centroids.
+   *
+   * @return A new  [[FitOperation]] to train the model using the training data set.
+   */
+  implicit def fitKMeans = {
+    new FitOperation[KMeans, Vector] {
+      override def fit(instance: KMeans, fitParameters: ParameterMap, trainingDS: DataSet[Vector])
+      : Unit = {
+        val resultingParameters = instance.parameters ++ fitParameters
+
+        // =================  INITIALIZATION OF KMEANS ==========================
+        val centroids: DataSet[Seq[LabeledVector]] = init(trainingDS, resultingParameters)
+
+        val numIterations: Int = resultingParameters.get(NumIterations).get
+
+        val finalCentroids = centroids.iterate(numIterations) { currentCentroids =>
+          val newCentroids: DataSet[LabeledVector] = trainingDS
+            .mapWithBcVariable(currentCentroids)
+              { (dataPoint, centroids) => selectNearestCentroid(dataPoint, centroids) }
+            .map(x => (x.label, x.vector, 1.0)).withForwardedFields("label->_1; vector->_2")
+            .groupBy(x => x._1)
+            .reduce((p1, p2) =>
+              (p1._1, (p1._2.asBreeze + p2._2.asBreeze).fromBreeze, p1._3 + p2._3))
+            // TODO replace addition of Breeze vectors by future build in flink function
+            .withForwardedFields("_1")
+            .map(x => {
+              BLAS.scal(1.0 / x._3, x._2)
+              LabeledVector(x._1, x._2)
+            })
+            .withForwardedFields("_1->label")
+
+          // currentCentroids contains only one element. So, this is output only once
+          currentCentroids.mapWithBcSet(newCentroids){
```

Comment:
If we store the centroids in a `DataSet[LabeledVector]` we wouldn't need to do that here.